### ISSUE TITLE
installing local dependencies from wheel/zip files

### ISSUE THREAD

**Issue by dustin-dawind** [basnijholt/unidep#211]
unidep doesn't seem to support installing local dependencies using a filepath to a .whl/.zip file, however, it is possible to install them directly via pip (e.g., `pip install path/example_package.whl`)

This would be extremely useful to me (and probably many others) who use hardware/instruments that have Python driver wrappers distributed as pre-compiled wheels or as zip files

On top of this, trying to use unidep to install via a local .whl/.zip produces very confusing error messages.

Here's an example traceback I've gotten:
```
Traceback (most recent call last):
  File "C:\miniforge3\envs\py312test\Scripts\unidep-script.py", line 9, in <module>
    sys.exit(_cli.main())
             ^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_cli.py", line 1455, in main
    _install_command(
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_cli.py", line 882, in _install_command
    requirements = parse_requirements(
                   ^^^^^^^^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_dependencies_parsing.py", line 415, in parse_requirements
    _update_data_structures(
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_dependencies_parsing.py", line 267, in _update_data_structures
    _add_local_dependencies(
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_dependencies_parsing.py", line 364, in _add_local_dependencies
    _update_data_structures(
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_dependencies_parsing.py", line 240, in _update_data_structures
    data = _load(path_with_extras.path, yaml, verbose=verbose)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\unidep\_dependencies_parsing.py", line 181, in _load
    return yaml.load(f)
           ^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\ruamel\yaml\main.py", line 449, in load
    constructor, parser = self.get_constructor_parser(stream)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\ruamel\yaml\main.py", line 500, in get_constructor_parser
    self.reader.stream = stream
    ^^^^^^^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\site-packages\ruamel\yaml\reader.py", line 120, in stream
    self.determine_encoding()
  File "C:\miniforge3\envs\py312test\Lib\site-packages\ruamel\yaml\reader.py", line 174, in determine_encoding
    self.update_raw()
  File "C:\miniforge3\envs\py312test\Lib\site-packages\ruamel\yaml\reader.py", line 263, in update_raw
    data = self.stream.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\miniforge3\envs\py312test\Lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 14: character maps to <undefined>
```

It appears that the `_load()` function in `_dependencies_parsing.py` is trying to load the file as a .yaml file, causing the decoding error.
In the meantime, it would seem beneficial to handle this error/use-case more explicitly - notifying the user that the current implementation only supports .toml/.yaml/.yml file extensions. 

Overall, though, I think unidep is an awesome package that simplifies dependency management immensely! I sincerely want to thank you for all the work you've put into this to get it to its current state

**Comment by basnijholt** [comment]
> Overall, though, I think unidep is an awesome package that simplifies dependency management immensely! I sincerely want to thank you for all the work you've put into this to get it to its current state

Thank you so much, that means a lot! 

What you are proposing makes a lot of sense and shouldn't be too hard to implement. 

I might be able to do it this week.

**Comment by basnijholt** [comment]
May I ask how you are using unidep? In which context?

**Comment by dustin-dawind** [comment]
I have unidep installed to a conda environment (created via mamba) and have the project dependencies listed within a pyproject.toml file at the root of my project. They're paths are listed relative to pyproject.toml

I am able to reproduce the issue using `unidep install .` at the project's root

I'm not sure if that fully answers your question or not, but let me know if you need any additional information and I'd be happy to provide you with whatever I'm able!

**Comment by basnijholt** [comment]
@dustin-dawind I am mostly just curious which features you use, not related to this specific issue per se!

Anyway I opened https://github.com/basnijholt/unidep/pull/212 which should implement the feature you asked for :tada:

**Comment by basnijholt** [comment]
I just released https://github.com/basnijholt/unidep/releases/tag/v0.64.0!

Please let me know whether it works for you 😄

**Comment by dustin-dawind** [comment]
@basnijholt Oh, yea I totally misunderstood what you were asking haha

I only just started using unidep the day I created this issue, so I'm probably not using a whole lot of unidep's capabilities. I work at a small photonics company as a Test Engineer and have recently been working on cleaning up our repo, making our code easier to maintain, and updating from Python 3.8 to Python 3.12.

I actually came across unidep because I was looking for a way to simplify our project's dependency maintainance/installation process. Since we have instruments that only distribute their Python wrappers on Pypi or even as directly downloadable wheels, we unfortunately can't solely use Conda for package management. Previously, my team was just directly dumping all Conda dependencies to a YAML file and then also maintaining a separate list of the Pip dependencies (both local and on Pypi). As you could probably already guess, this makes it extremely difficult to know which packages are direct dependencies and which are dependencies of dependencies - not to mention it obviously also makes it very difficult to know which packages to update (so much so, that my team had basically never updated anything since the original environment was first created!)

So in short, I'm using unidep mainly to simplify managing our dependency list and reduce the amount of commands needed to install everything to new environments

**Comment by basnijholt** [comment]
Thanks a lot for sharing! This is basically exactly the same reason I created unidep 😄 

We've found its ability to create conda lock-files also very valuable to get 100% reproducible environments.

**Comment by dustin-dawind** [comment]
> I just released https://github.com/basnijholt/unidep/releases/tag/v0.64.0!
> 
> Please let me know whether it works for you 😄

Had a few days off work, so only getting back to this now, but thanks for the quick upload!

I've ran the updated version on my end and I'm unfortunately running into a different error now

```
raise ParserSyntaxError(
    packaging._tokenizer.ParserSyntaxError: Expected end or semicolon (after URL and whitespace)
        <wheel name>.whl @ file://C:/Users/<username>/<path to local package>/<wheel name>.whl
```

From what I can tell, this is due to my account username unfortunately having a space in it (this was set up for me by a 3rd party IT company before I was hired, so unfortunately was out of my control). Is there anything I can do on my end to work around this?

**Comment by basnijholt** [comment]
Would you be able to share an example of the `pyproject.toml`?

You can of course "anonymize" it but leave the structure the same.

**Comment by dustin-dawind** [comment]
Sure thing! This is the exact structure of my `pyproject.toml`:

```
[build-system]
requires = [
    "setuptools",
    "unidep"
]
build-backend = "setuptools.build_meta"

[project]
name = "[package_name]"
dynamic = [
    "version",
    "dependencies"
]

requires-python = ">=3.12"

maintainers = [
    { name = "[name1]", email = "[email1]" },
    { name = "[name2]", email = "[email2]" },
    { name = "[name3]", email = "[email3]" },
    { name = "[name4]", email = "[email4]" },
    { name = "[name5]", email = "[email5]" },
]

[tool.unidep]
channels = ["conda-forge"]
dependencies = [
    { conda = "[package1]>=[version]" },
    { conda = "[package2]>=[version]" },
    { pip   = "[package3]"},
    { conda = "[package4]>=[version]" },
    { conda = "[package5]>=[version]" },
    { conda = "[package6]>=[version]" },
    { pip   = "[package7]" },
    { conda = "[package8]>=[version]" },
    { conda = "[package9]>=[version]" },
    { conda = "[package10]>=[version]" },
    { conda = "[package11]>=[version]" },
    { conda = "[package12]>=[version]" },
    { conda = "[package13]>=[version]" },
    { conda = "[package14]>=[version]" },
    { conda = "[package15]>=[version]" },
    { pip   = "[package16]" },
    { conda = "[package17]=[version]" },
    { conda = "[package18]>=[version]" },
    { conda = "[package19]>=[version]" },
    { conda = "[package20]>=[version]" },
    { conda = "[package21]>=[version]" },
    { pip   = "[package22]" },
    { pip   = "[package23]" },
    { conda = "[package24]>=[version]" },
    { conda = "[package25]>=[version]" },
    { conda = "[package26]>=[version]" },
    { conda = "[package27]>=[version]" },
    { conda = "[package28]=[version]" },
    { conda = "[package29]>=[version]" },
    { pip   = "[package30]" },
    { conda = "[package31]>=[version]" },
    { conda = "[package32]>=[version]" },
    { conda = "[package33]>=[version]" },
    { conda = "[package34]>=[version]" },
    { conda = "[package35]>=[version]" },
    { conda = "[package36]>=[version]" },
    { conda = "[package37]>=[version]" },
    { conda = "[package38]>=[version]" },
    { conda = "[package39]>=[version]" },
]

local_dependencies  = [
#    "../[repo1]/[path]/[local zip].zip",
    "../[repo1]/[path]/[local whl].whl",
#    "../[repo2]/[path]/[local editable package that uses setup.py]"
]
```

**Comment by basnijholt** [comment]
I can locally reproduce the problem with:
```diff
diff --git a/example/setup_py_project/requirements.yaml b/example/setup_py_project/requirements.yaml
index 49582fc..4e34e6d 100644
--- a/example/setup_py_project/requirements.yaml
+++ b/example/setup_py_project/requirements.yaml
@@ -12,6 +12,7 @@ dependencies:
   - pip: aiokef
 local_dependencies:
   - ../setuptools_project  # depends on setuptools_project
+  - ./my folder/pipefunc-0.46.0-py3-none-any.whl
 platforms:
   - linux-64
   - osx-64
```

And creating a folder `my folder` with the whl.

This results in:

```bash
  error in setup_py_project setup command: 'install_requires' must be a string or iterable of strings containing valid project/version requirement specifiers; Expected end or semicolon (after URL and whitespace)
      pipefunc-0.46.0-py3-none-any.whl @ file:///Users/basnijholt/Code/unidep/example/setup_py_project/my folder/pipefunc-0.46.0-py3-none-any.whl
                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /Users/basnijholt/micromamba/envs/unidep/bin/python3.13 /Users/basnijholt/micromamba/envs/unidep/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpvjdfz7uj
  cwd: /Users/basnijholt/Code/unidep/example/setup_py_project
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Traceback (most recent call last):
  File "/Users/basnijholt/micromamba/envs/unidep/bin/unidep", line 8, in <module>
    sys.exit(_cli.main())
             ~~~~~~~~~^^
  File "/Users/basnijholt/Code/unidep/unidep/_cli.py", line 1455, in main
    _install_command(
    ~~~~~~~~~~~~~~~~^
        *(args.files or [Path()]),
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<13 lines>...
        verbose=args.verbose,
        ^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/basnijholt/Code/unidep/unidep/_cli.py", line 999, in _install_command
    _pip_install_local(
    ~~~~~~~~~~~~~~~~~~^
        *sorted(installable),
        ^^^^^^^^^^^^^^^^^^^^^
    ...<4 lines>...
        conda_run=conda_run,
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/basnijholt/Code/unidep/unidep/_cli.py", line 860, in _pip_install_local
    subprocess.run(pip_command, check=True)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/basnijholt/micromamba/envs/unidep/lib/python3.13/subprocess.py", line 577, in run
    raise CalledProcessError(retcode, process.args,
                             output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['micromamba', 'run', '--prefix', '/Users/basnijholt/micromamba/envs/unidep', '/Users/basnijholt/micromamba/envs/unidep/bin/python3.13', '-m', 'pip', 'install', '--no-dependencies', '--verbose', '/Users/basnijholt/Code/unidep/example/hatch_project', '/Users/basnijholt/Code/unidep/example/setup_py_project/my folder/pipefunc-0.46.0-py3-none-any.whl', '/Users/basnijholt/Code/unidep/example/setuptools_project', './example/setup_py_project']' returned non-zero exit status 1.
```

**Comment by dustin-dawind** [comment]
Glad it's not just me haha

I know that when directly using pip with filepaths containing spaces, you can surround the filepath with quotes to get it to handle the filepath correctly. Not sure if that's helpful in this case, but thought I'd mention it just in case

**Comment by basnijholt** [comment]
Still debugging this, just leaving some notes.

```
import subprocess

subprocess.run(
    [
        "pip",
        "install",
        "--no-dependencies",
        "--verbose",
        "-e",
        "./example/setup_py_project",
        "/Users/basnijholt/Code/unidep/example/setup_py_project/my folder/pipefunc-0.46.0-py3-none-any.whl",
    ],
    check=True,
)
```
Reproduces the error, however, removing 
```
        "-e",
        "./example/setup_py_project",
```

works without issues...

**Comment by basnijholt** [comment]
@dustin-dawind, I believe this is a pip bug https://github.com/pypa/pip/issues/13139.

Hopefully there is a workaround.

**Comment by basnijholt** [comment]
Fix now in [v0.65.0](https://github.com/basnijholt/unidep/releases/tag/v0.65.0) :tada:

Please let me know if it works now!

**Comment by basnijholt** [comment]
BTW re:
```toml
dependencies = [
    { conda = "[package1]>=[version]" },
    { conda = "[package2]>=[version]" },
    { pip   = "[package3]"},
    { conda = "[package4]>=[version]" },
```
If dependencies have the same name on conda and pip you can just do:
```toml
dependencies = [
    "[package1]>=[version]" ,
    "[package2]>=[version]" ,
    { pip   = "[package3]"},  # only on pip
    { conda = "[package4]>=[version]" },  # only on conda
```

**Comment by dustin-dawind** [comment]
I haven't been able to *fully* verify everything just yet because there is some lingering scratch code in our repo that I need to find and purge, but as far as I can tell everything is working great now! Thank you so much for all your work on this!

I've told all my coworkers about unidep, and hopefully soon we will start porting all our other repos over to using it too!
